File: backend/app/schemas/notifications.py
from datetime import datetime
from typing import Any, Dict, List, Optional
from pydantic import BaseModel, EmailStr, Field


class NotificationResponse(BaseModel):
    id: int
    type: str
    title: str
    message: str
    action_url: Optional[str] = None
    metadata_json: Dict[str, Any] = Field(default_factory=dict)
    is_read: bool
    created_at: datetime
    read_at: Optional[datetime] = None

    model_config = {"from_attributes": True}


class NotificationSettingsUpdate(BaseModel):
    email_notifications: bool = True
    resume_ready: bool = True
    job_alerts: bool = True
    weekly_report: bool = True
    interview_reminder: bool = True
    security_alerts: bool = True
    marketing_emails: bool = False
    application_updates: bool = True


class NotificationSettingsResponse(NotificationSettingsUpdate):
    id: int
    user_id: int
    updated_at: datetime

    model_config = {"from_attributes": True}


class ApplicationSettingsUpdate(BaseModel):
    mode: str = Field(default="manual", pattern="^(manual|automatic)$")
    auto_apply_enabled: bool = False
    minimum_match_score: float = Field(default=85, ge=0, le=100)
    preferred_companies: List[str] = Field(default_factory=list)
    preferred_locations: List[str] = Field(default_factory=list)
    salary_range: Dict[str, Any] = Field(default_factory=dict)
    experience_range: Dict[str, Any] = Field(default_factory=dict)
    remote_only: bool = False
    exclude_companies: List[str] = Field(default_factory=list)
    maximum_daily_applications: int = Field(default=5, ge=0, le=100)
    working_hours: Dict[str, Any] = Field(default_factory=dict)
    daily_job_search_enabled: bool = False
    daily_job_search_time: str = "09:00"
    daily_job_search_platforms: List[str] = Field(default_factory=list)
    jobs_per_platform: int = Field(default=20, ge=1, le=50)


class ApplicationSettingsResponse(ApplicationSettingsUpdate):
    id: int
    user_id: int
    updated_at: datetime

    model_config = {"from_attributes": True}


class EmailLogResponse(BaseModel):
    id: int
    to_email: str
    subject: str
    template_name: Optional[str] = None
    provider: str
    status: str
    attempts: int
    error_message: Optional[str] = None
    sent_at: Optional[datetime] = None
    created_at: datetime

    model_config = {"from_attributes": True}


class TestEmailRequest(BaseModel):
    to_email: Optional[EmailStr] = None
    template_name: str = "welcome"


class ForgotPasswordRequest(BaseModel):
    email: EmailStr


class ResetPasswordRequest(BaseModel):
    token: str
    new_password: str = Field(min_length=6)


class VerifyEmailRequest(BaseModel):
    token: str


class LimitsResponse(BaseModel):
    platform: str
    used: int
    limit: int
    remaining: int
    reset_at: datetime


class PlatformLimitUpdate(BaseModel):
    daily_search_limit: int = Field(default=20, ge=0, le=500)
    daily_application_limit: int = Field(default=5, ge=0, le=100)
    is_enabled: bool = True

File: frontend/services/notifications.ts
import { apiFetch } from "./api";

export interface NotificationSettings {
  email_notifications: boolean;
  resume_ready: boolean;
  job_alerts: boolean;
  weekly_report: boolean;
  interview_reminder: boolean;
  security_alerts: boolean;
  marketing_emails: boolean;
  application_updates: boolean;
}

export interface ApplicationSettings {
  mode: "manual" | "automatic";
  auto_apply_enabled: boolean;
  minimum_match_score: number;
  preferred_companies: string[];
  preferred_locations: string[];
  salary_range: Record<string, unknown>;
  experience_range: Record<string, unknown>;
  remote_only: boolean;
  exclude_companies: string[];
  maximum_daily_applications: number;
  working_hours: Record<string, unknown>;
  daily_job_search_enabled: boolean;
  daily_job_search_time: string;
  daily_job_search_platforms: string[];
  jobs_per_platform: number;
}

export interface NotificationItem {
  id: number;
  type: string;
  title: string;
  message: string;
  action_url?: string | null;
  is_read: boolean;
  created_at: string;
}

export interface EmailLogItem {
  id: number;
  to_email: string;
  subject: string;
  template_name?: string | null;
  provider: string;
  status: string;
  attempts: number;
  error_message?: string | null;
  sent_at?: string | null;
  created_at: string;
}

export interface UsageLimitItem {
  platform: string;
  used: number;
  limit: number;
  remaining: number;
  reset_at: string;
}

export function fetchNotificationSettings() {
  return apiFetch<NotificationSettings>("/api/settings/notifications");
}

export function updateNotificationSettings(payload: NotificationSettings) {
  return apiFetch<NotificationSettings>("/api/settings/notifications", {
    method: "POST",
    body: payload,
  });
}

export function fetchApplicationSettings() {
  return apiFetch<ApplicationSettings>("/api/settings/application-mode");
}

export function updateApplicationSettings(payload: ApplicationSettings) {
  return apiFetch<ApplicationSettings>("/api/settings/application-mode", {
    method: "POST",
    body: payload,
  });
}

export function fetchNotifications() {
  return apiFetch<NotificationItem[]>("/api/notifications");
}

export function fetchEmailHistory() {
  return apiFetch<EmailLogItem[]>("/api/email/history");
}

export function sendTestEmail(template_name = "welcome") {
  return apiFetch<{ success: boolean; sent: boolean; to_email: string }>("/api/email/test", {
    method: "POST",
    body: { template_name },
  });
}

export function fetchUsageLimits() {
  return apiFetch<UsageLimitItem[]>("/api/limits");
}
